quantile/cm/heap.go

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cm
+
+type minHeap []float64
+
+// Len returns the number of values in the heap
+func (h minHeap) Len() int { return len(h) }
+
+// Min returns the minimum value from the heap
+func (h minHeap) Min() float64 { return h[0] }
+
+// Push pushes a value onto the heap
+func (h *minHeap) Push(value float64) {
+	*h = append(*h, value)
+	h.shiftUp(h.Len() - 1)
+}
+
+// Pop pops the minimum value from the heap
+func (h *minHeap) Pop() float64 {
+	var (
+		old = *h
+		n   = old.Len()
+		val = old[0]
+	)
+
+	old[0], old[n-1] = old[n-1], old[0]
+	h.heapify(0, n-1)
+	*h = (*h)[0 : n-1]
+	return val
+}
+
+func (h minHeap) shiftUp(i int) {
+	for {
+		parent := (i - 1) / 2
+		if parent == i || h[parent] <= h[i] {
+			break
+		}
+		h[parent], h[i] = h[i], h[parent]
+		i = parent
+	}
+}
+
+func (h minHeap) heapify(i, n int) {
```

Comment:
Any reason not to use the stdlib heap and just supply the correct `Less` logic to make it a min heap? https://golang.org/pkg/container/heap/